Adds new eslint rule to detect package-relative imports

scinos opened the pull request:
### Change

This PR introduces a new eslint rule to forbid package-relative imports. The rule is disabled for now (see Plan).
This PR also introduces a new dependency on `eslint-nibble`, that we'll use to selectively fix eslint rules.


### Background

Calypso allows package-relative imports, like

```js
import config from 'config';
import userFactory from 'lib/user';
```

While this saves us from having to type an endless stream of `../../../../../` for relative imports, it has some drawbacks: it makes the code harder to statically analyze, and such relative import has to be configured in all our tools that do module resolution (node.js, webpack, typescript, IDEs...).

An alternative would be to use regular NPM imports using `wp-calypso`:
```js
import config from 'wp-calypso/config';
import userFactory from 'wp-calypso/lib/user';
```

This works because `./client` is actually an NPM package called `wp-calypso`, and it is referenced as a dependency in the root's `package.json`. 

#### Migration plan

After playing with some other ideas (see section 'Alternatives'), I think the best way to do this migration is an eslint rule. We can turn it on to prevent new files to use the package-relative import, and we can use https://www.npmjs.com/package/eslint-nibble to gradually fix parts of our codebase.

#### Alternatives

Ideas I consider and didn't work:

* Use a bunch of `bash`+`sed`+`grep` magic: it works for simple cases, but super hard to do when the import is split into several lines. Plus it will require extra work to forbid new files to use package-relative imports.
* Use `calypso-codemods`: the wrapper around `jscodeshift` doesn't allow configuration of the rule or change the parser to support TypeScript. Plus it will require extra work to forbid new files to use package-relative imports.

So the idea is:

1. Introduce a new eslint rule (this PR) and blog about it.
2. Convert small pieces of our codebase to use relative imports, ensure everything works:
   * [x] Compiling evergreen/fallback
   * [x] Starting Node server
   * [x] Running tests
   * [x] Compiling TS projects
   * [ ] ...
3. Enable the rule to prevent new files to use the wrong syntax.
4. Keep migrating the whole codebase.

### Testing instructions

You can run the rule for a subset of files by:
1. Open `.eslintrc` and configure the rule `wpcalypso/no-package-relative-imports` from `off` to `error`.
2. Run `./node_modules/.bin/eslint-nibble --ext .js,.jsx,.ts,.tsx --rule wpcalypso/no-package-relative-imports <path>` (warning: applying it in all `./client` takes several minutes)
3. Apply the fixes and validate that the results make sense.

scinos committed:
Adds new eslint rule to detect package-relative imports

scinos committed:
Fix typo in rule name

sgomes commented on packages/eslint-plugin-wpcalypso/lib/rules/no-package-relative-imports.js:
```diff
@@ -0,0 +1,111 @@
+		const fileName = file.name;
+		if ( file.isDirectory() ) return fileName;
+		const extension = path.extname( fileName );
+		if ( extension !== '.js' ) return fileName;
```
We probably need to account for TypeScript `.ts` files as well, and possibly `.json` files too.

scinos replied:
Good point.
I'll make it configurable, as each project may have their own set of optional extensions.

sgomes replied:
That's an excellent idea 👍

dmsnell replied:
what about `.tsx` and `.jsx`?

sgomes commented on packages/eslint-plugin-wpcalypso/lib/rules/no-package-relative-imports.js:
```diff
@@ -0,0 +1,111 @@
+};
+
+const reportIfPackageRelative = ( { context, node, dependency } ) => {
+	const { mapping, warnOnDynamicImport } = context.options[ 0 ];
```
These options don't appear to be documented. That's not a blocking concern for this PR, considering the early state of this package, but something to consider.

sgomes commented on packages/eslint-plugin-wpcalypso/lib/rules/no-package-relative-imports.js:
```diff
@@ -0,0 +1,111 @@
+	const { value: dependencyName, type: dependencyType } = dependency;
+
+	if ( dependencyType !== 'Literal' ) {
+		if ( warnOnDynamicImport ) {
```
Dynamic import usually refers to the async `import( 'modulename' )` syntax which returns a promise, regardless of whether you pass in a literal or a variable. I believe this piece of code is referring to a the latter case specifically, though, where the module name being passed in is not a literal, but rather a variable or an expression of some sort.

If I'm right, perhaps "dynamic import" isn't the best name for this option, since it could lead to confusion? I'm struggling to think of a better alternative, however. Maybe just be literal about it, and call this a "non-literal import"?

scinos replied:
Agree about the name, it is confusing. Eslint calls "dynamic require" to CJS require with an expression (as opposed to a literal), the same meaning I'm using here: https://github.com/benmosher/eslint-plugin-import/blob/HEAD/docs/rules/no-dynamic-require.md

I'll change it to something else.

dmsnell replied:
Not sure about this but I believe `webpack` has stringent rules on what level of dynamism is possible; they still have to be able to resolve the name, meaning you have to be able to provide a glob or some expression in `require.ensure()` (or its modern equivalent)

by this nature though we can't go inspecting the code to reverse-engineer the value used and I support the decision to ignore any of these. if we're not using literals as the import name it's a can of worms that should be caught by PR review anyway

scinos replied:
For completeness, I only found two cases of this in our codebase:

* https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/step-components.js#L71
* https://github.com/Automattic/wp-calypso/blob/master/client/server/bundler/sections-loader.js#L29

saramarcondes replied:
Would "programmatic import name" work?

scinos commented on .eslintrc.js:
```diff
@@ -223,5 +224,18 @@ module.exports = {
+			{
+				mapping: [
+					{
+						dir: path.join( __dirname, 'client' ),
```
This means: For every import like `import * from 'X'`, when X is a subdirectory of the path specified in `dir`, replace it with `wp-calypso/X`

scinos committed:
Add Configuration section to rule README

sgomes reviewed (commented):
Thanks for working on this, @scinos! I l really like the idea of getting rid of this ambiguity 👍 

I added a few comments below before a more complete pass later today.

scinos committed:
Add comments to the rule

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@scinos please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/826729). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

dmsnell commented on packages/eslint-plugin-wpcalypso/lib/rules/no-package-relative-imports.js:
```diff
@@ -0,0 +1,128 @@
+	mapping.forEach( ( { dir, module } ) => {
+		const deps = getRelativeImports( dir );
+
+		if ( deps.includes( dependencyName.split( '/' )[ 0 ] ) ) {
```
would we ant to use any kind of path resolution here to avoid rare-but-possible cases like `./././././thing/`

scinos replied:
Added a check of imports starting with `.` or `/` following Node's resolution algorithm. If it starts with any of those characters, then the import is ignored. If not, it doesn't really matter if it is something crazy like `foo/../../bar` (is that even valid?), it will be replaced with `wp-calypso/foo/../../bar` if `./client/foo` exists.

dmsnell reviewed (commented):
Pretty good work @scinos - this should not only catch issues but I think it will help people who aren't aware of what they should be doing to be able to find out.

saramarcondes commented on packages/eslint-plugin-wpcalypso/docs/rules/no-package-relative-imports.md:
```diff
@@ -0,0 +1,58 @@
+
+Which works because `wp-calypso` is a valid package in our repository, points to `./client` and is declared as a dependency in the root `package.json`. This will work out of the box in all module resolution systems.
+
+This rule forbid the former approach and can auto-fix it to the latter.
```
```suggestion
This rule forbids the former approach and can auto-fix it to the latter.
```

saramarcondes commented on packages/eslint-plugin-wpcalypso/docs/rules/no-package-relative-imports.md:
````diff
@@ -0,0 +1,58 @@
+const config = require('config');
+```
+
+The following patterns are not warnings:
````
```suggestion
The following patterns are not errors:
```

sirreal replied:
`error` or `warning` can be configured. [Eslint docs](https://eslint.org/docs/rules/for-direction) talk about "correct" and "incorrect" code for a given rule.

saramarcondes commented on .eslintrc.js:
```diff
@@ -223,5 +224,18 @@ module.exports = {
 		],
+		// Disabled for now until we finish the migration
+		'wpcalypso/no-package-relative-imports': [
+			'off',
```
Why not at least `warn`?

scinos replied:
I'd like to do more testing before enabling the rule in any form, to make sure people don't accidentally start adding `wp-calypso/` everywhere before we are 100% sure it works in all cases.

scinos committed:
Allow passing extensions in the rule config

scinos committed:
Do not enable no-package-relative-imports rule

sgomes reviewed (approved):
LGTM, great work @scinos! Thanks for taking care of all the suggestions.

The rule works very nicely, and so does the auto-fixing 👍

scinos commented:
Given that this got an initial pass of reviews, I addressed most of the feedback and the rule is not enabled yet, I'm going to merge it. We can always change it before enabling it.
